### ISSUE TITLE
Dependency Injection in Startup.cs

### ISSUE THREAD

**Issue by KevinMallinson** [sendgrid/sendgrid-csharp#880]
Hi, sorry for not following the template, but regarding issues #847  and #521, I was wondering if there is any progress?

I would like to inject a SendGridClient through DI to a controller, instead of instantiating a new client every request.

I can't rely on `services.AddHttpClient<ISendGridClient, SendGridClient>();` because I need to add the `Key` that is typically called in the SendGridClient constructor.

Is this workflow being looked in to?

Thanks

**Comment by thinkingserious** [comment]
Hello @KevinMallinson,

I don't know the answer to this offhand, I will need to investigate further.

In the mean time, could you please elaborate on what the behavior you are seeking would look like in code if we had already implemented it so that I may better understand your workflow? 

With Best Regards,

Elmer

**Comment by KevinMallinson** [comment]
The behaviour would look something like this (psuedocode):

`MySendGridClient.cs`
```
public class MySendGridClient : SendGridClient
{
    public class MySendGridClient(HttpClient httpClient /*The client we will keep active */, string apiKey)
    {
        //Add custom headers, etc
        // ...

        //Add API Key - call parent method
        base.AddApiKey(apiKey);
    }
}


`Startup.cs`
```
public void ConfigureServices(IServiceCollection services)
{
    services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_1);

    services.AddHttpClient<SendGridClient, MySendGridClient>();
}

`MyController.cs`
```
public class MyController
{
    public MyController(SendGridClient myClient)
    {
        // myClient is passed via dependency injection
        // it is always instantiated, no need to re-instantiate each time
    }
}
```

Hopefully that makes sense, 

Thanks

**Comment by thinkingserious** [comment]
Thank you @KevinMallinson,

Your description was very helpful, thanks for taking the time!

I've added this to our internal backlog for consideration as well as opened it up for community contribution.

With Best Regards,

Elmer

**Comment by aevitas** [comment]
This could be added relatively easily by adding a constructor that accepts an `IOptions<SendGridClientOptions>` argument, unwraps the option type, and calls the existing constructor with the `SendGridClientOptions options` signature.

In your DI configuration, you'd do something along the lines of:
```
services.Configure<SendGridClientOptions>(Configuration.GetSection("SendGrid");
services.AddScoped<ISendGridClient, SendGridClient>();
```

Let me know if you want me to apply these changes, I'll happily implement them and submit a PR.

**Comment by thinkingserious** [comment]
Hello @aevitas,

A PR would be most welcome. Thank you!

With best regards,

Elmer

**Comment by akunzai** [comment]
Since Sendgrid [v9.12.0](https://github.com/sendgrid/sendgrid-csharp/releases/tag/v9.12.0) released.

My workaround is create a service extension method

```csharp
using System;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection.Extensions;
using Microsoft.Extensions.Options;

namespace SendGrid.Extensions
{
    public static class ServiceCollectionExtensions
    {
        public static IHttpClientBuilder AddSendGrid(this IServiceCollection services,
            Action<IServiceProvider, SendGridClientOptions> configureOptions)
        {
            services.TryAddSingleton(resolver => Options.Create(resolver.GetRequiredService<IOptions<SendGridClientOptions>>().Value));

            services.AddOptions<SendGridClientOptions>().Configure<IServiceProvider>((options, resolver) => configureOptions(resolver, options))
                .PostConfigure(options =>
                {
                    // validation
                    if (string.IsNullOrWhiteSpace(options.ApiKey))
                    {
                        throw new ArgumentNullException(nameof(options.ApiKey));
                    }
                });

            services.TryAddTransient<ISendGridClient>(resolver => resolver.GetRequiredService<SendGridClient>());

            return services.AddHttpClient<SendGridClient>();
        }
    }
}
```

Usage

```csharp
private void ConfigureServices(IServiceCollection services){
    services.AddSendGrid((resolver, options) =>
    {
        var configuration = resolver.GetService<IConfiguration>();
        options.ApiKey = configuration["SendGrid:ApiKey"];
    });
}
```

**Comment by KevinMallinson** [comment]
@aevitas @thinkingserious 

#839 was merged, will #904 require changes, and is this still planned on being merged?

**Comment by masaab** [comment]
You can use Install-Package SendGrid.Extensions.DependencyInjection -Version 1.0.0
And use it in your startup Configure method. Something like this.

 builder.Services.AddSendGrid(options => {
                options.ApiKey = Environment.GetEnvironmentVariable("SENDGRID_API_KEY");
